micro-rdk/src/common/conn/viam.rs

```diff
@@ -390,6 +391,85 @@ where
         }
     }
 }
+
+pin_project_lite::pin_project! {
+    struct SystemTask<A, S> {
+        app_client_tasks: A,
+        server_task: S,
+        force_shutdown_wait_time: Duration,
+        timer: Option<Timer>,
+    }
+}
+
+impl<A, S> SystemTask<A, S> {
+    fn new(
+        server_task: S,
+        app_client_tasks: A,
+        force_shutdown_wait_time: Duration,
+    ) -> Pin<Box<Self>> {
+        Box::pin(SystemTask {
+            app_client_tasks,
+            server_task,
+            force_shutdown_wait_time,
+            timer: None,
+        })
+    }
+}
+
+impl<A, S> Future for SystemTask<A, S>
+where
+    A: Future<Output = Result<(), errors::ServerError>> + std::marker::Unpin,
+    S: Future<Output = Result<(), errors::ServerError>> + std::marker::Unpin,
+{
+    type Output = (A::Output, S::Output);
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        let pinned = self.project();
+        if shutdown_requested() && pinned.timer.is_none() {
+            let dur = pinned.force_shutdown_wait_time;
+            *pinned.timer = Some(Timer::after(*dur));
+            return Poll::Pending;
+        }
+
+        if let Some(timer) = pinned.timer {
+            match timer.poll(cx) {
+                Poll::Pending => match pinned.app_client_tasks.poll(cx) {
+                    Poll::Pending => Poll::Pending,
+                    Poll::Ready(app_client_task_state) => {
+                        log::info!("app client tasks gracefully finished");
+                        match pinned.server_task.poll(cx) {
+                            Poll::Pending => Poll::Pending,
+                            Poll::Ready(server_task_state) => {
+                                Poll::Ready((app_client_task_state, server_task_state))
+                            }
+                        }
+                    }
+                },
+                Poll::Ready(_) => {
+                    let app_tasks_res = match pinned.app_client_tasks.poll(cx) {
+                        Poll::Pending => Err(errors::ServerError::ServerTaskShutdownTimeout),
```

Comment:
shouldn't this be `AppTasksShutdownTimeout`?

Reply:
I didn't feel like making a separate error for each one, but I would be open to renaming if you think it's confusing